spec/fastsync/fastsync_concurrency_with_peers.tla

```diff
@@ -0,0 +1,543 @@
+    or do nothing 
+*)
+ScheduleRoutine ==
+    \* handle the incoming message on demuxSchedule
+    \/ HandleDemuxSchedule
+    \* do nothing
+    \/ /\ UNCHANGED buffers
+       /\ UNCHANGED peer
+
+(* 
+SendRoutine --
+    either handle the message sent by the demux routine on the demuxSend buffer,
+    or do nothing 
+*)
+SendRoutine ==
+    \* handle the incoming message on demuxSend
+    \/ HandleDemuxSend
+    \* do nothing
+    \/ /\ UNCHANGED buffers
+       /\ UNCHANGED peer
+      
+(*
+Peer --
+    either non-deterministically send peer messages to the receive routine
+    or do noting
+*)       
+Peer ==
+    \* send peer messages
+    \/ SendPeerMsgs
+    \* do nothing 
+    \/ /\ UNCHANGED buffers
+       /\ UNCHANGED peer
+
+Buffer ==
+    \/ /\ ~Ready(receiveDemux)
+       /\ ~Full(receiveDemux)
+       /\ receiveDemux' = BufRcv(receiveDemux)
+       /\ UNCHANGED <<demuxSend, demuxProcess, processDemux, demuxSchedule, scheduleDemux, peerIncoming, peerOutgoing, peerTurn>>
+    \/ /\ ~Ready(demuxSend)
+       /\ ~Full(demuxSend)
+       /\ demuxSend' = BufRcv(demuxSend)
+       /\ UNCHANGED <<receiveDemux, demuxProcess, processDemux, demuxSchedule, scheduleDemux, peerIncoming, peerOutgoing, peerTurn>>
+    \/ /\ ~Ready(demuxProcess)
+       /\ ~Full(demuxProcess)
+       /\ demuxProcess' = BufRcv(demuxProcess)
+       /\ UNCHANGED <<receiveDemux, demuxSend, processDemux, demuxSchedule, scheduleDemux, peerIncoming, peerOutgoing, peerTurn>>
+    \/ /\ ~Ready(processDemux)
+       /\ ~Full(processDemux)
+       /\ processDemux' = BufRcv(processDemux)
+       /\ UNCHANGED <<receiveDemux, demuxSend, demuxProcess, demuxSchedule, scheduleDemux, peerIncoming, peerOutgoing, peerTurn>>
+    \/ /\ ~Ready(demuxSchedule)
+       /\ ~Full(demuxSchedule)
+       /\ demuxSchedule' = BufRcv(demuxSchedule)
+       /\ UNCHANGED <<receiveDemux, demuxSend, demuxProcess, processDemux, scheduleDemux, peerIncoming, peerOutgoing, peerTurn>>
+    \/ /\ ~Ready(scheduleDemux)
+       /\ ~Full(scheduleDemux)
+       /\ scheduleDemux' = BufRcv(scheduleDemux)
+       /\ UNCHANGED <<receiveDemux, demuxSend, demuxProcess, processDemux, demuxSchedule, peerIncoming, peerOutgoing, peerTurn>>   
+    \/ /\ \E peerID \in PeerIDs : /\ ~Ready(peerIncoming[peerID])
+                                  /\ ~Full(peerIncoming[peerID])
+                                  /\ peerIncoming' = [peerIncoming EXCEPT ![peerID] = BufRcv(@)]
+       /\ UNCHANGED <<receiveDemux, demuxSend, demuxProcess, processDemux, demuxSchedule, scheduleDemux, peerOutgoing, peerTurn>>
+    \/ /\ \E peerID \in PeerIDs : /\ ~Ready(peerOutgoing[peerID])
+                                  /\ ~Full(peerOutgoing[peerID])
+                                  /\ peerOutgoing' = [peerOutgoing EXCEPT ![peerID] = BufRcv(@)]
+       /\ UNCHANGED <<receiveDemux, demuxSend, demuxProcess, processDemux, demuxSchedule, scheduleDemux, peerIncoming, peerTurn>>       
```

Comment:
Should we separate internal communication from external in separate operators

Reply:
Not sure what you mean. This is an internal operation of the peerOutgoing buffers (now renamed to peerToReceive), where the inMsg is added to the queue if the buffer is not ready and not full.